panic when executeBatchJobs meet error

### What did you do?

make a slow downstrem and decrease sycner's maxDMLConnectionTimeout
so every execSQL will meet a timeout error

---
```
[2022/01/13 14:33:19.604 +08:00] [ERROR] [db.go:221] ["execute statements failed after retry"] [task=test] [unit="binlog replication"] [queries="[DELETE FROM `sspanel`.`sspanel_user` WHERE `id` = ? LIMIT 1 REPLACE INTO `sspanel`.`sspanel_user` (`id`,`password`,`last_login`,`is_superuser`,`username`,`first_name`,`last_name`,`email`,`is_staff`,`is_active`,`date_joined`,`inviter_id`,`balance`,`invitecode_num`,`level`,`level_expire_time`,`theme`,`vmess_uuid`,`download_traffic`,`last_use_time`,`ss_password`,`ss_port`,`total_traffic`,`upload_traffic`,`uid`) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)]"] [arguments="[[7] [7 ehco1 <nil> 1 1 1 1 1 1 1 2022-01-11 09:56:23.000000 1 <nil> 1 1 2022-01-11 09:56:38.000000 d 1 0 <nil> 1 1 1 1 <nil>]]"] [error="[code=10006:class=database:scope=not-set:level=high], Message: execute statement failed: REPLACE INTO `sspanel`.`sspanel_user` (`id`,`password`,`last_login`,`is_superuser`,`username`,`first_name`,`last_name`,`email`,`is_staff`,`is_active`,`date_joined`,`inviter_id`,`balance`,`invitecode_num`,`level`,`level_expire_time`,`theme`,`vmess_uuid`,`download_traffic`,`last_use_time`,`ss_password`,`ss_port`,`total_traffic`,`upload_traffic`,`uid`) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?), RawCause: invalid connection"]
panic: runtime error: index out of range [1] with length 1

goroutine 750 [running]:
github.com/pingcap/tiflow/dm/syncer.(*DMLWorker).executeBatchJobs.func1(0x14005947e98, 0x14005dbebb0, 0x0, 0x14005947ea8, {0x14004c79710, 0x1, 0x1}, 0x14005947e40)
	github.com/pingcap/tiflow/dm/syncer/dml_worker.go:209 +0xb4
github.com/pingcap/tiflow/dm/syncer.(*DMLWorker).executeBatchJobs(0x14005dbebb0, 0x0, {0x14004c79710, 0x1, 0x1})
	github.com/pingcap/tiflow/dm/syncer/dml_worker.go:248 +0x3dc
github.com/pingcap/tiflow/dm/syncer.(*DMLWorker).executeJobs(0x14005dbebb0, 0x0, 0x14005e39f80)
	github.com/pingcap/tiflow/dm/syncer/dml_worker.go:180 +0x330
created by github.com/pingcap/tiflow/dm/syncer.(*DMLWorker).run
	github.com/pingcap/tiflow/dm/syncer/dml_worker.go:102 +0xac
```

## RootCause

https://github.com/pingcap/tiflow/blob/0cd880f6887f4626ba1516e8b48b7b72a9c87233/dm/syncer/dml_worker.go#L209

use wrong job index `affect` 
should use `affect-1` and check `jobs` length beacuse `affect` always equal  to job count in one batch
https://github.com/pingcap/tiflow/blob/0cd880f6887f4626ba1516e8b48b7b72a9c87233/dm/pkg/conn/baseconn.go#L154

### What did you expect to see?

task paused but worker still alive

### What did you see instead?

worker panic

### Versions of the cluster

master

### current status of DM cluster (execute `query-status <task-name>` in dmctl)

```console
(paste current status of DM cluster here)
```

--- a/dm/syncer/dml_worker.go
+++ b/dm/syncer/dml_worker.go
@@ -196,17 +196,24 @@ func (w *DMLWorker) executeJobs(queueID int, jobCh chan *job) {
 // executeBatchJobs execute jobs with batch size.
 func (w *DMLWorker) executeBatchJobs(queueID int, jobs []*job) {
 	var (
-		affect int
-		db     = w.toDBConns[queueID]
-		err    error
-		dmls   = make([]*DML, 0, len(jobs))
+		affect  int
+		queries []string
+		args    [][]interface{}
+		db      = w.toDBConns[queueID]
+		err     error
+		dmls    = make([]*DML, 0, len(jobs))
 	)
 
 	defer func() {
 		if err == nil {
 			w.successFunc(queueID, len(dmls), jobs)
 		} else {
-			w.fatalFunc(jobs[affect], err)
+			if len(queries) == len(jobs) {
+				w.fatalFunc(jobs[affect], err)
+			} else {
+				w.logger.Warn("length of queries not equals length of jobs, cannot determine which job failed, use first one instead", zap.Int("queries", len(queries)), zap.Int("jobs", len(jobs)))
+				w.fatalFunc(jobs[0], err)
+			}
 		}
 	}()
 
@@ -230,7 +237,7 @@ func (w *DMLWorker) executeBatchJobs(queueID int, jobs []*job) {
 	for _, j := range jobs {
 		dmls = append(dmls, j.dml)
 	}
-	queries, args := w.genSQLs(dmls)
+	queries, args = w.genSQLs(dmls)
 	failpoint.Inject("WaitUserCancel", func(v failpoint.Value) {
 		t := v.(int)
 		time.Sleep(time.Duration(t) * time.Second)

--- a/dm/syncer/dml_worker.go
+++ b/dm/syncer/dml_worker.go
@@ -247,8 +247,9 @@ func (w *DMLWorker) executeBatchJobs(queueID int, jobs []*job) {
 	})
 
 	failpoint.Inject("ErrorOnLastDML", func(_ failpoint.Value) {
-		if len(dmls) > len(jobs) {
-			affect, err = len(dmls), terror.ErrDBExecuteFailed.Delegate(errors.New("ErrorOnLastDML"), "mock")
+		if len(queries) > len(jobs) {
+			w.logger.Error("error on last queries", zap.Int("queries", len(queries)), zap.Int("jobs", len(jobs)))
+			affect, err = len(queries)-1, terror.ErrDBExecuteFailed.Delegate(errors.New("ErrorOnLastDML"), "mock")
 		}
 	})
 }

--- a/dm/syncer/dml_worker.go
+++ b/dm/syncer/dml_worker.go
@@ -245,6 +245,12 @@ func (w *DMLWorker) executeBatchJobs(queueID int, jobs []*job) {
 			affect, err = 0, terror.ErrDBExecuteFailed.Delegate(errors.New("SafeModeExit"), "mock")
 		}
 	})
+
+	failpoint.Inject("ErrorOnLastDML", func(_ failpoint.Value) {
+		if len(dmls) > len(jobs) {
+			affect, err = len(dmls), terror.ErrDBExecuteFailed.Delegate(errors.New("ErrorOnLastDML"), "mock")
+		}
+	})
 }
 
 // genSQLs generate SQLs in single row mode or multiple rows mode.

--- a/dm/syncer/dml_worker.go
+++ b/dm/syncer/dml_worker.go
@@ -211,8 +211,12 @@ func (w *DMLWorker) executeBatchJobs(queueID int, jobs []*job) {
 			if len(queries) == len(jobs) {
 				w.fatalFunc(jobs[affect], err)
 			} else {
-				w.logger.Warn("length of queries not equals length of jobs, cannot determine which job failed, use first one instead", zap.Int("queries", len(queries)), zap.Int("jobs", len(jobs)))
-				w.fatalFunc(jobs[0], err)
+				w.logger.Warn("length of queries not equals length of jobs, cannot determine which job failed", zap.Int("queries", len(queries)), zap.Int("jobs", len(jobs)))
+				newJob := job{
+					startLocation:   jobs[0].startLocation,
+					currentLocation: jobs[len(jobs)-1].currentLocation,
+				}
+				w.fatalFunc(&newJob, err)
 			}
 		}
 	}()

--- a/dm/tests/shardddl1/run.sh
+++ b/dm/tests/shardddl1/run.sh
@@ -784,9 +784,10 @@ function DM_CAUSALITY_USE_DOWNSTREAM_SCHEMA() {
 }
 
 function DM_DML_EXECUTE_ERROR_CASE() {
-	run_sql_source1 "insert into ${shardddl1}.${tb1}(a,b,c) values(1,1,1)"
-	run_sql_source1 "update ${shardddl1}.${tb1} set a=a+1, b=b+1, c=c+1 where a=1"
+	run_sql_source1 "insert into ${shardddl1}.${tb1}(a,b) values(1,1)"
+	run_sql_source1 "update ${shardddl1}.${tb1} set b=b+1 where a=1"
 	check_sync_diff $WORK_DIR $cur/conf/diff_config.toml
+	read v1
 }
 
 function DM_DML_EXECUTE_ERROR() {
@@ -801,14 +802,13 @@ function DM_DML_EXECUTE_ERROR() {
 	check_rpc_alive $cur/../bin/check_worker_online 127.0.0.1:$WORKER2_PORT
 
 	run_case DML_EXECUTE_ERROR "single-source-no-sharding" \
-		"run_sql_source1 \"create table ${shardddl1}.${tb1} (a int primary key, b int unique, c int);\"" \
+		"run_sql_source1 \"create table ${shardddl1}.${tb1} (a int primary key, b int);\"" \
 		"clean_table" ""
 }
 
 function run() {
 	init_cluster
 	init_database
-
 
 	DM_DML_EXECUTE_ERROR
 	start=1

--- a/dm/tests/shardddl1/run.sh
+++ b/dm/tests/shardddl1/run.sh
@@ -783,6 +783,28 @@ function DM_CAUSALITY_USE_DOWNSTREAM_SCHEMA() {
 		"clean_table" ""
 }
 
+function DM_DML_EXECUTE_ERROR_CASE() {
+	run_sql_source1 "insert into ${shardddl1}.${tb1}(a,b,c) values(1,1,1)"
+	run_sql_source1 "update ${shardddl1}.${tb1} set a=a+1, b=b+1, c=c+1 where a=1"
+	check_sync_diff $WORK_DIR $cur/conf/diff_config.toml
+}
+
+function DM_DML_EXECUTE_ERROR() {
+	# mock downstream has a high latency and upstream has a high workload
+	ps aux | grep dm-worker | awk '{print $2}' | xargs kill || true
+	check_port_offline $WORKER1_PORT 20
+	check_port_offline $WORKER2_PORT 20
+	export GO_FAILPOINTS='github.com/pingcap/tiflow/dm/syncer/ErrorOnLastDML'
+	run_dm_worker $WORK_DIR/worker1 $WORKER1_PORT $cur/conf/dm-worker1.toml
+	run_dm_worker $WORK_DIR/worker2 $WORKER2_PORT $cur/conf/dm-worker2.toml
+	check_rpc_alive $cur/../bin/check_worker_online 127.0.0.1:$WORKER1_PORT
+	check_rpc_alive $cur/../bin/check_worker_online 127.0.0.1:$WORKER2_PORT
+
+	run_case COMPACT "single-source-no-sharding" \
+		"run_sql_source1 \"create table ${shardddl1}.${tb1} (a int primary key, b int unique, c int);\"" \
+		"clean_table" ""
+}
+
 function run() {
 	init_cluster
 	init_database
@@ -799,6 +821,7 @@ function run() {
 	DM_RestartMaster
 	DM_ADD_DROP_COLUMNS
 	DM_COLUMN_INDEX
+	DM_DML_EXECUTE_ERROR
 	start=1
 	end=5
 	for i in $(seq -f "%03g" ${start} ${end}); do

--- a/dm/tests/shardddl1/run.sh
+++ b/dm/tests/shardddl1/run.sh
@@ -786,12 +786,15 @@ function DM_CAUSALITY_USE_DOWNSTREAM_SCHEMA() {
 function DM_DML_EXECUTE_ERROR_CASE() {
 	run_sql_source1 "insert into ${shardddl1}.${tb1}(a,b) values(1,1)"
 	run_sql_source1 "update ${shardddl1}.${tb1} set b=b+1 where a=1"
-	check_sync_diff $WORK_DIR $cur/conf/diff_config.toml
-	read v1
+
+	check_log_contain_with_retry "length of queries not equals length of jobs" $WORK_DIR/worker1/log/dm-worker.log $WORK_DIR/worker2/log/dm-worker.log
+	run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+		"query-status test" \
+		"\"RawCause\": \"ErrorOnLastDML\"" 1 \
+		"Paused" 1
 }
 
 function DM_DML_EXECUTE_ERROR() {
-	# mock downstream has a high latency and upstream has a high workload
 	ps aux | grep dm-worker | awk '{print $2}' | xargs kill || true
 	check_port_offline $WORKER1_PORT 20
 	check_port_offline $WORKER2_PORT 20
@@ -810,6 +813,18 @@ function run() {
 	init_cluster
 	init_database
 
+	DM_COMPACT
+	DM_COMPACT_USE_DOWNSTREAM_SCHEMA
+	DM_MULTIPLE_ROWS
+	DM_CAUSALITY
+	DM_CAUSALITY_USE_DOWNSTREAM_SCHEMA
+	DM_UpdateBARule
+	DM_RENAME_TABLE
+	DM_RENAME_COLUMN_OPTIMISTIC
+	DM_RemoveLock
+	DM_RestartMaster
+	DM_ADD_DROP_COLUMNS
+	DM_COLUMN_INDEX
 	DM_DML_EXECUTE_ERROR
 	start=1
 	end=5

--- a/dm/tests/shardddl1/run.sh
+++ b/dm/tests/shardddl1/run.sh
@@ -794,13 +794,13 @@ function DM_DML_EXECUTE_ERROR() {
 	ps aux | grep dm-worker | awk '{print $2}' | xargs kill || true
 	check_port_offline $WORKER1_PORT 20
 	check_port_offline $WORKER2_PORT 20
-	export GO_FAILPOINTS='github.com/pingcap/tiflow/dm/syncer/ErrorOnLastDML'
+	export GO_FAILPOINTS='github.com/pingcap/tiflow/dm/syncer/ErrorOnLastDML=return()'
 	run_dm_worker $WORK_DIR/worker1 $WORKER1_PORT $cur/conf/dm-worker1.toml
 	run_dm_worker $WORK_DIR/worker2 $WORKER2_PORT $cur/conf/dm-worker2.toml
 	check_rpc_alive $cur/../bin/check_worker_online 127.0.0.1:$WORKER1_PORT
 	check_rpc_alive $cur/../bin/check_worker_online 127.0.0.1:$WORKER2_PORT
 
-	run_case COMPACT "single-source-no-sharding" \
+	run_case DML_EXECUTE_ERROR "single-source-no-sharding" \
 		"run_sql_source1 \"create table ${shardddl1}.${tb1} (a int primary key, b int unique, c int);\"" \
 		"clean_table" ""
 }
@@ -809,18 +809,7 @@ function run() {
 	init_cluster
 	init_database
 
-	DM_COMPACT
-	DM_COMPACT_USE_DOWNSTREAM_SCHEMA
-	DM_MULTIPLE_ROWS
-	DM_CAUSALITY
-	DM_CAUSALITY_USE_DOWNSTREAM_SCHEMA
-	DM_UpdateBARule
-	DM_RENAME_TABLE
-	DM_RENAME_COLUMN_OPTIMISTIC
-	DM_RemoveLock
-	DM_RestartMaster
-	DM_ADD_DROP_COLUMNS
-	DM_COLUMN_INDEX
+
 	DM_DML_EXECUTE_ERROR
 	start=1
 	end=5

